source/sigv4.c

```diff
@@ -1814,10 +1819,69 @@ static void generateCredentialScope( const SigV4Parameters_t * pSigV4Params,
 
 /*-----------------------------------------------------------*/
 
-    static SigV4Status_t setQueryStringFieldsAndValues( const char * pQuery,
-                                                        size_t queryLen,
-                                                        size_t * pNumberOfParameters,
-                                                        CanonicalContext_t * pCanonicalRequest )
+    static void processAmpersantInQueryString( bool fieldHasValue,
+                                               size_t currQueryIndex,
+                                               size_t startOfFieldOrValue,
+                                               size_t * pCurrParamCount,
+                                               const char * pQuery,
+                                               CanonicalContext_t * pCanonicalRequest )
+    {
+        bool previousParamIsValid = true;
+
+        assert( pCurrParamCount != NULL );
+        assert( pQuery != NULL );
+        assert( pCanonicalRequest != NULL );
+
+        /* Ignore unnecessary '&' that represent empty parameter names.
+         * Trimmable '&'s can be leading, trailing or repeated as separators between parameter
+         * pairs.
+         * For example, this function will prune "&p1=v1&&p2=v2&" to "p1=v1&p2=v2". */
+        if( !( fieldHasValue ) && ( ( ( currQueryIndex - startOfFieldOrValue ) == 0U ) ) )
```

Comment:
If there's a trimmable `&`, would `startOfFieldOrValue` need to be incremented?

Reply:
The `startOfFieldOrValue` value is always incremented in the calling code of this function. I will move that logic to this function to easily notice that.

Reply:
Didn't notice that that was changed. It looks fine as is then, as it looks like `startOfFieldOrValue` should always be incremented when encountering an `&`, especially for the case of "&&" which wasn't handled previously.